RecoTracker/DeDx/plugins/SiPixelTrackProbQXYProducer.cc

```diff
@@ -0,0 +1,187 @@
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  int factorial(int);
+
+  // ----------member data ---------------------------
+  edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> tTopoToken_;
+  edm::EDGetTokenT<reco::TrackCollection> trackToken_;
+};
+
+using namespace reco;
+using namespace std;
+using namespace edm;
+
+SiPixelTrackProbQXYProducer::SiPixelTrackProbQXYProducer(const edm::ParameterSet& iConfig) : tTopoToken_(esConsumes()) {
+  produces<reco::SiPixelTrackProbQXYCollection>();
+  produces<reco::SiPixelTrackProbQXYAssociation>();
+  trackToken_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("tracks"));
+}
+
+void SiPixelTrackProbQXYProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<reco::TrackCollection> trackCollectionHandle;
+  iEvent.getByToken(trackToken_, trackCollectionHandle);
+  const TrackCollection& trackCollection(*trackCollectionHandle.product());
+  float probQonTrack = 0.0;
+  float probXYonTrack = 0.0;
+  float probQonTrackNoL1 = 0.0;
+  float probXYonTrackNoL1 = 0.0;
+  // creates the output collection
+  auto resultSiPixelTrackProbQXYColl = std::make_unique<reco::SiPixelTrackProbQXYCollection>();
+  std::vector<int> indices;
+
+  // Loop through the tracks
+  for (unsigned int j = 0; j < trackCollection.size(); j++) {
+    const reco::Track& track = trackCollection[j];
+
+    int numRecHits = 0;
+    int numRecHitsNoL1 = 0;
+    float probQonTrackWMulti = 1;
+    float probXYonTrackWMulti = 1;
+    float probQonTrackWMultiNoL1 = 1;
+    float probXYonTrackWMultiNoL1 = 1;
+
+    // Loop through the rechits on the given track
+    auto hb = track.recHitsBegin();
+    for (unsigned int h = 0; h < track.recHitsSize(); h++) {
+      auto recHit = *(hb + h);
+      const SiPixelRecHit* pixhit = dynamic_cast<const SiPixelRecHit*>(recHit);
+      if (pixhit == nullptr)
+        continue;
+      if (!pixhit->isValid())
+        continue;
+      if (pixhit->geographicalId().det() != DetId::Tracker)
+        continue;
+
+      //Retrieve tracker topology from geometry
+      const TrackerTopology* const tTopo = &iSetup.getData(tTopoToken_);
+
+      // Have a separate variable that excludes Layer 1
+      // Layer 1 was very noisy in 2017/2018
+      float probQNoL1 = 0;
+      float probXYNoL1 = 0;
+      if (tTopo->pxbLayer(pixhit->geographicalId()) != 1) {
+        probQNoL1 = pixhit->probabilityQ();
+        probXYNoL1 = pixhit->probabilityXY();
+        if (probQNoL1 != 0) {  // only save the non-zero rechits
+          numRecHitsNoL1++;
+          // Calculate alpha term needed for the combination
+          probQonTrackWMultiNoL1 *= probQNoL1;
+          probXYonTrackWMultiNoL1 *= probXYNoL1;
+        }
+      }
+
+      // Have a variable that includes all layers and disks
+      float probQ = 0;
+      float probXY = 0;
+      probQ = pixhit->probabilityQ();
+      probXY = pixhit->probabilityXY();
```

Comment:
```suggestion
      float probQ = pixhit->probabilityQ();
      float probXY = pixhit->probabilityXY();
```
?
seems more economical.

Reply:
I was thinking about doing this, but then I was thinking what if probabilityQ() doesnt return anything? In that case wouldnt it be better to do the prob=0 part? If you think it's safe to do the suggestion, I'm happy to do that.

Reply:
How can `probabilityQ()` not return something? You mean like it throws? in that case you have worse problems to deal with.